lib/basedriver/driver.js

```diff
@@ -195,33 +210,41 @@ class BaseDriver extends MobileJsonWireProtocol {
-    let nextCommand = this.curCommand.then(() => { // eslint-disable-line promise/prefer-await-to-then
-      // if we unexpectedly shut down, we need to reject every command in
-      // the queue before we actually try to run it
+    let res;
+    if (this.isCommandsQueueEnabled) {
+      // What we're doing here is pretty clever. this.curCommand is always
+      // a promise representing the command currently being executed by the
+      // driver, or the last command executed by the driver (it starts off as
+      // essentially a pre-resolved promise). When a command comes in, we tack it
+      // to the end of this.curCommand, essentially saying we want to execute it
+      // whenever this.curCommand is done. We call this new promise nextCommand,
+      // and its resolution is what we ultimately will return to whomever called
+      // us. Meanwhile, we reset this.curCommand to _be_ nextCommand (but
+      // ignoring any rejections), so that if another command comes into the
+      // server, it gets tacked on to the end of nextCommand. Thus we create
+      // a chain of promises that acts as a queue with single concurrency.
+      let nextCommand = this.curCommand.then(() => { // eslint-disable-line promise/prefer-await-to-then
+        // if we unexpectedly shut down, we need to reject every command in
+        // the queue before we actually try to run it
+        if (this.shutdownUnexpectedly) {
+          return B.reject(new errors.NoSuchDriverError('The driver was unexpectedly shut down!'));
+        }
+        // We also need to turn the command into a cancellable promise so if we
+        // have an unexpected shutdown event, for example, we can cancel it from
+        // outside, rejecting the current command immediately
+        this.curCommandCancellable = new B((r) => { r(); }).then(() => { // eslint-disable-line promise/prefer-await-to-then
+          return this[cmd](...args);
+        }).cancellable();
+        return this.curCommandCancellable;
+      });
+      this.curCommand = nextCommand.catch(() => {});
+      res = await nextCommand;
+    } else {
       if (this.shutdownUnexpectedly) {
-        return B.reject(new errors.NoSuchDriverError('The driver was unexpectedly shut down!'));
+        throw new errors.NoSuchDriverError('The driver was unexpectedly shut down!');
       }
-      // We also need to turn the command into a cancellable promise so if we
-      // have an unexpected shutdown event, for example, we can cancel it from
-      // outside, rejecting the current command immediately
-      this.curCommandCancellable = new B((r) => { r(); }).then(() => { // eslint-disable-line promise/prefer-await-to-then
-        return this[cmd](...args);
-      }).cancellable();
-      return this.curCommandCancellable;
-    });
-    this.curCommand = nextCommand.catch(() => {});
-    let res = await nextCommand;
+      res = await B.resolve(() => this[cmd](...args));
```

Comment:
if you say _await_ is smart enough to process this then B.resolve can be removed